### ISSUE TITLE
[css-syntax] Dashed-ident rules and error recovery

### ISSUE THREAD

**Issue by andruud** [w3c/csswg-drafts#9336]
After a recent edit in relation to relaxed nesting, [css-syntax now says this](https://drafts.csswg.org/css-syntax/#consume-qualified-rule):

> [<{-token>](https://drafts.csswg.org/css-syntax/#tokendef-open-curly)
If the first two non-[<whitespace-token>](https://drafts.csswg.org/css-syntax/#typedef-whitespace-token) values of rule’s prelude are an [<ident-token>](https://drafts.csswg.org/css-syntax/#typedef-ident-token) whose value starts with "--" followed by a [<colon-token>](https://drafts.csswg.org/css-syntax/#typedef-colon-token), [consume the remnants of a bad declaration](https://drafts.csswg.org/css-syntax/#consume-the-remnants-of-a-bad-declaration) from input, with nested, and return nothing.

Unless I'm mistaken, this means that encountering this situation top-level means we'll treat the entire rest of the stylesheet as a "bad declaration".

```
--foo:hover {
  color: red;
}
/* Lots of other innocent and valid rules here */
```
That is probably a bit extreme, and makes the change a bit hard to ship, since it _could_ affects existing sites (drastically).

Perhaps we can limit the behavior to when `nested` is `true`? Or something else? @tabatkins

**Comment by tabatkins** [comment]
Yup, this looks like just a mistake; I didn't mean for this hungry behavior to apply at the top-level.

The *intention* of the hungry behavior is that it's already *very* difficult for custom properties to *actually* ever be invalid, so an "ambiguous rule" like `--foo:hover {...}` would *virtually never* fail the "looks like a declaration" test and actually parse as a rule. To avoid that being a super rare case, I just catch it in the "consume a rule" algo and force it to continue to act like a custom property.

But since custom props (or any props at all) can't exist on the top level of a stylesheet, there's no reason to worry about it. That is, `--really-a-prop:hover and-more-weird-values` is just unreasonable to interpret as a property at the top level of a stylesheet.

So yeah, I propose that we do indeed guard against this. Specifically, in <https://drafts.csswg.org/css-syntax/#consume-a-qualified-rule>, in the `{-token` branch, if we see something that looks like a custom property, then we behave differently based on whether it's nested or not: if nested, we do the current behavior (force it to be a custom property and throw it away), if not nested, we just consume the {} block and then throw it away.

This maintains the constant I want to maintain about where `--foo:hover {...}` rules are allowed (nowhere), and avoids eating an entire stylesheet.

Agenda+ to confirm this with the WG.

**Comment by cdoublev** [comment]
I do not see in which nested context `--foo:hover { color: blue; } not-semicolon` would not be *consumed* as a (valid) declaration.

So I do not understand why checking `nesting` is needed (in [*consume a qualified rule*](https://drafts.csswg.org/css-syntax-3/#consume-qualified-rule), cf. previous comment) instead of consuming `{}` *indifferently* whether the consumed rule is nested or not.

---

(I'm trying to guess from here, if that helps explain it to me what I am missing.)

From the note titled *What's this check for* below the *consume a qualified rule* algo (and also in the previous comment):

  > This ensures that `--foo:hover { color: blue; }` is consistently invalid everywhere [...]

Does it mean that this is consistently an invalid *rule* everywhere?

  > Authors can write a custom property that takes a `{}-block` in its value, even combined with other thing; if that custom property is then invalid (due to an invalidly-written `var()` function, for example), [...]

Do you mean that `--foo: var(<not-custom-property-name>)` should fail to be *consumed* as a declaration?

[2.2 Error Handling](https://drafts.csswg.org/css-syntax-3/#error-handling) says that the grammar must be checked after *each* construct is parsed, but it is non-normative and it seems inefficient to retry parsing a declaration as a rule because its value were invalid according to the grammar.

I may have misuderstood https://github.com/w3c/csswg-drafts/issues/8834#issuecomment-1677876075, ie. what it means to check that a declaration or rule is *valid in the context* (which [I think](https://github.com/w3c/csswg-drafts/issues/8834#issuecomment-1685266923) is not required anymore, btw).

**Comment by tabatkins** [comment]
`--foo:hover { color: blue; } var(1);` will fail to parse as a valid declaration, and instead attempt to be parsed as a qualified rule.

> So I do not understand why checking nesting is needed

In a nested context, a `--foo: ...` is *probably* a declaration, so consuming the whole thing to next the same-level semicolon is likely to be the correct behavior.

At the top level, properties aren't allowed; if you've written `--foo:...` it's much more likely you are trying to write a rule. You don't see many stray `color: blue;` declarations at the top level of a stylesheet, after all. So we should consume just a rule-worth of tokens; that's more likely to match what the author intended. Plus, since declarations aren't valid at the top level, it's very unlikely that a top-level semicolon will appear *at all*, so a stray `--foo:...` would likely consume *the entire rest of the stylesheet* if it was treated as a declaration.

> Do you mean that `--foo: var(<not-custom-property-name>)` should fail to be consumed as a declaration?

Correct. That's step 9 of ["consume a declaration"](https://drafts.csswg.org/css-syntax-3/#consume-a-declaration)

> it seems inefficient to retry parsing a declaration as a rule because its value was invalid according to the grammar.

Sure, it would be inefficient. But you don't actually have to try, you just have to act identically to an impl that *did* try. The vast majority of the time you can just quit immediately, or at least very quickly. See the impl notes at the end of ["consume a block's contents"](https://drafts.csswg.org/css-syntax-3/#consume-a-blocks-contents), which offers guidance about this exact topic.

**Comment by tabatkins** [comment]
> ACTION: TabAtkins to look into using a more generic term than 'nested' for the flag for mixed declaration+rule contexts

Closing the loop on this action:

Most of the "consume" algorithms use the "nested" flag. In almost all cases it really *does* mean "am I nested in a {} block", because it's used to determine whether a `}` stops the parsing immediately (because it's the closing brace of a parent block) or not.

The only exception is in "consume a qualified rule": in addition to the above-described usage, it's also used to decide how to handle a custom property. In theory this is referencing the "mixed declarations and rules, or just rules" decision, but in practice it's to make a custom property showing up at the top level of a stylesheet work better (so it doesn't consume the entire rest of the stylesheet), so it's *actually* caring about nesting too.

So, I'm concluding that "nested" is still an appropriate name for this flag, and this action needs no further change.
